src/inmanta/server/services/compilerservice.py

```diff
@@ -1068,3 +1068,20 @@ async def environment_action_deleted(self, env: model.Environment) -> None:
     async def recalculate_queue_count_cache(self) -> None:
         async with self._queue_count_cache_lock:
             self._queue_count_cache = await data.Compile.get_total_length_of_all_compile_queues()
+
+    async def cancel_compile(self, env_id: uuid.UUID) -> None:
+        """
+        Cancel the ongoing compile for the given environment (if any). After the cancellation, this method
+        will not start the execution of a new compile, if a compile request would be present in the compile
+        queue for the given environment.
+
+        :param env_id: The id of the environment for which the compile should be cancelled.
+        """
+        compile_task: Optional[Task[None | Result]] = self._recompiles.pop(env_id, None)
+        if compile_task:
+            compile_task.cancel()
```

Comment:
We'll have to carefully consider the potential race between the task finishing and this call. e.g. if the task finishes, we have to make sure it does not start the next compile while we only cancel this one.

I'm not going deeper until the other thread is resolved.

Reply:
We always first halt the environment and then call into `cancel_compile()`. No new compiles are started on halted environment so this race condition shouldn't exist.